### PR TITLE
Use Openthread Hardware Entropy Collector

### PR DESCRIPTION
**Pull request by luwannest**

### CODE REVIEW

**Comment by mszczodrak** [comment]
LGTM 👍

File: third_party/mbedtls/Makefile.am [hunk 1]
```diff
@@ -39,6 +39,7 @@ libmbedcrypto_a_CPPFLAGS                      = \
 
 libmbedcrypto_a_SOURCES                       = \
     mbedcrypto.c                                \
+    mbedentropy.c                               \
```
Comment (jwhui) [inline third_party/mbedtls/Makefile.am]: Can we include this in `mbedcrypto.c`?

File: third_party/mbedtls/mbedentropy.c [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include <platform/random.h>
+
+int mbedtls_hardware_poll( void *data,
+                    unsigned char *output, size_t len, size_t *olen )
+{
+    (void)data;
+
+    if (otPlatSecureRandomGet((uint16_t)len, (uint8_t *)output, (uint16_t *)olen) != kThreadError_None)
+    {
+        return -1;
```
Comment (jwhui) [inline third_party/mbedtls/mbedentropy.c]: Should return `MBEDTLS_ERR_ENTROPY_SOURCE_FAILED`.

File: third_party/mbedtls/mbedentropy.c [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ *    limitations under the License.
+ */
+
+#include <stddef.h>
+
+#include <openthread-types.h>
+#include <platform/random.h>
+
+int mbedtls_hardware_poll( void *data,
+                    unsigned char *output, size_t len, size_t *olen )
```
Comment (jwhui) [inline third_party/mbedtls/mbedentropy.c]: I know `astyle` does not descend into `third-party` yet.  But let's try to keep the code style consistent.

**Comment by luwannest** [comment]
@jwhui , one problem that encountered after including mbedtls_hardware_entropy in embedcrypto.c is the linking order dependency.
libmbedcrypto.a depends on libopenthread-posix.a, meanwhile libopenthread-posix.a depends on libmbedcrypto.a.
Just like the previous patch, creating a new mbedentropy.c. That depends on libopenthread-posix.a, but libopenthread-posix.a does not depend on it.
Or do we have any other better ways?

**Comment by jwhui** [comment]
@luwannest, that makes sense.  FYI, I'm currently working on changes to OT's use of mbedtls.  You can view that pending change at 097e4ec.

**Comment by luwannest** [comment]
@jwhui , could you take a look at this code? This is needed after enable commissioning.

**Comment by chshu** [comment]
LGMT:+1:

**Comment by jwhui** [comment]
Thanks for updating this.  We recently updated the mbedtls bindings to all be placed in `src/core/crypto` rather than in `third_party`.  Is it possible to add this binding to `mbedtls.cpp` instead?

**Comment by luwannest** [comment]
The main reason I put in third_party/mbedtls is that mbedtls_hardware_entropy calls OT API to get the entropy source.
To my understanding of core/crypto, it is to use mbedtls, but not be used by mbedtls.

Not sure it makes sense or not. If not, I will try to move it.

**Comment by jwhui** [comment]
I understand your reasoning.  However, in general, any original code contributed to the OpenThread project should live in `src/`.

**Comment by luwannest** [comment]
@jwhui , I encountered the library dependency order problem while trying to move mbedtls_entropy_poll to crypto folder.
If I move mbedtls_entropy_poll to crypto folder, it will be in libopenthread.a, but it will be used by libmbedcrypto.a. And here I can't change the order,because libopenthread.a depends on libmbedcrypto.a.

One option is to put mbedtls_entropy_poll in crypto folder, but build as a standalone archive, and put it behind libmbedcrypto.a and before libopenthread-posix.a (and any other hardware platforms).
But this seems not to be consistent with current build system, that src/core will be the libopenthread.a.

Could you give me any comments on this?

**Comment by jwhui** [comment]
Okay, I see what you mean.  In the interest of moving this forward, let's keep the mapping in `third_party/mbedtls` as you had it before.  Thanks.

**Comment by luwannest** [comment]
Will figure out a way to move this to src later, if this is ok only for now.

**Comment by jwhui** [comment]
LGTM 👍

**Pull request merged**